test/unit/fact_cleaner_test.rb

```diff
@@ -48,6 +48,48 @@ class FactCleanerTest < ActiveSupport::TestCase
     assert_include remaining_of(root_6_fact), root_6_fact
   end
 
+  test "it cleans excluded facts" do
+    cleaner.stubs(:excluded_facts).returns(['ignored*', '*bad'])
+    cleaner.stubs(:delete_orphaned_facts).returns(0)
+    good_fact_name = FactoryBot.create(:fact_name, :name => 'good_fact')
+    ignored_fact_name = FactoryBot.create(:fact_name, :name => 'ignored01')
+    bad_fact_name = FactoryBot.create(:fact_name, :name => 'empty_bad')
+    FactoryBot.create(:fact_value, :fact_name => good_fact_name)
+    FactoryBot.create(:fact_value, :fact_name => ignored_fact_name)
+    cleaner.clean!
+    assert_equal 2, cleaner.deleted_count
+    assert_include remaining_of(good_fact_name, ignored_fact_name, bad_fact_name), good_fact_name
+  end
+
+  test 'it cleans hierarchical facts' do
+    cleaner.stubs(:excluded_facts).returns(['ignored'])
+    cleaner.stubs(:delete_orphaned_facts).returns(0)
+
+    ignored_facts = [
+      FactoryBot.create(:fact_name, :name => 'ignored'),
+      FactoryBot.create(:fact_name, :name => 'ignored::something'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored::something')
+    ]
+
+    good_facts = [
+      FactoryBot.create(:fact_name, :name => 'not_ignored'),
+      FactoryBot.create(:fact_name, :name => 'not_ignored::something'),
+      FactoryBot.create(:fact_name, :name => 'common::not_ignored'),
+      FactoryBot.create(:fact_name, :name => 'common::not_ignored::something'),
+      FactoryBot.create(:fact_name, :name => 'ignored_shouldnt'),
+      FactoryBot.create(:fact_name, :name => 'ignored_shouldnt::something'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored_shouldnt'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored_shouldnt::something')
+    ]
```

Comment:
I must insist on real world facts, because (I haven't tested but) I believe your code does not work for real facts. Please include the following real facts I just captured from my server:

```
interfaces => em1,lo
ipaddress6_em1 => 2620:52:0:800:2247:ffff:ffff:95a0
ipaddress_em1 => 10.8.0.1
macaddress_em1 => 20:47:47:89:95:bb
mtu_em1 => 1500
netmask_em1 => 255.255.224.0
network_em1 => 10.8.0.0
dhcp_servers::em1 => 10.10.10.11
```

Again, I haven't tested myself, but I think your mathing code won't match most of those.